addon/src/mocks/mock-request.js

```diff
@@ -139,55 +151,94 @@ export default class {
     return this;
   }
 
-  paramsMatch(request) {
+  // eslint-disable-next-line no-unused-vars
+  paramsMatch({ request, params }) {
+    const requestUrl = new URL(request.url, window.location.origin);
     if (!isEmptyObject(this.someQueryParams)) {
-      return isPartOf(
-        toParams(request.queryParams),
-        toParams(this.someQueryParams),
-      );
+      // request must have every specified item in someQueryParams, but can have more.
+      const someQueryParams = new URLSearchParams(param(this.someQueryParams)); // convert json -> URLSearchParams
+      return someQueryParams
+        .keys()
+        .every((key) =>
+          isEqual(
+            someQueryParams.getAll(key),
+            requestUrl.searchParams.getAll(key),
+          ),
+        );
     }
     if (!isEmptyObject(this.queryParams)) {
-      return isEquivalent(
-        toParams(request.queryParams),
-        toParams(this.queryParams),
+      // request must have exactly specified queryParams, no more no less.
+      const queryParams = new URLSearchParams(param(this.queryParams));
+      return (
+        queryParams
+          .keys()
+          .every((key) =>
+            isEqual(
+              queryParams.getAll(key),
+              requestUrl.searchParams.getAll(key),
+            ),
+          ) &&
+        requestUrl.searchParams
+          .keys()
+          .every((key) =>
+            isEqual(
+              queryParams.getAll(key),
+              requestUrl.searchParams.getAll(key),
+            ),
+          )
       );
     }
     return true;
   }
 
   attributesMatch(requestBody, matchParams) {
-    if (isEmptyObject(matchParams)) {
-      return true;
+    return isMatch(requestBody, matchParams);
+  }
+
+  async getRequestBody(request) {
+    // Try to figure out request body type without relying on request headers
+    try {
+      return await request.clone().json();
+    } catch (e) {
+      // continue
+    }
+    try {
+      return await request.clone().formData();
+    } catch (e) {
+      // continue
+    }
+    try {
+      return await request.clone().text();
+    } catch (e) {
+      // continue
```

Comment:
we can't really rely on request headers being set in order to determine this, so we just try each one.